internal/staging/stdscript/scriptv0.go

```diff
@@ -714,3 +714,122 @@ func MultiSigScriptV0(threshold int, pubKeys ...[]byte) ([]byte, error) {
 
 	return builder.Script()
 }
+
+// AtomicSwapDataPushesV0 houses the data pushes found in hash-based atomic swap
+// contracts using version 0 scripts.
+type AtomicSwapDataPushesV0 struct {
+	RecipientHash160 [20]byte
+	RefundHash160    [20]byte
+	SecretHash       [32]byte
+	SecretSize       int64
+	LockTime         int64
+}
+
+// ExtractAtomicSwapDataPushesV0 returns the data pushes from an atomic swap
+// contract using version 0 scripts if it is one.  It will return nil otherwise.
+//
+// NOTE: Atomic swaps are not considered standard script types by the dcrd
+// mempool policy and should be used with P2SH.  The atomic swap format is also
+// expected to change to use a more secure hash function in the future.
+func ExtractAtomicSwapDataPushesV0(redeemScript []byte) *AtomicSwapDataPushesV0 {
+	// Local constants for convenience.
+	const (
+		maxMathOpCodeLen = txscript.MathOpCodeMaxScriptNumLen
+		maxCltvLen       = txscript.CltvMaxScriptNumLen
+	)
+
+	// An atomic swap is of the form:
+	//  IF
+	//   SIZE <secret size> EQUALVERIFY SHA256 <32-byte secret> EQUALVERIFY DUP
+	//   HASH160 <20-byte recipient hash>
+	//  ELSE
+	//   <locktime> CHECKLOCKTIMEVERIFY DROP DUP HASH160 <20-byte refund hash>
+	//  ENDIF
+	//  EQUALVERIFY CHECKSIG
+	type templateMatch struct {
+		expectCanonicalInt bool
+		maxIntBytes        int
+		opcode             byte
+		extractedInt       int64
+		extractedData      []byte
+	}
+	var template = [20]templateMatch{
+		{opcode: txscript.OP_IF},
+		{opcode: txscript.OP_SIZE},
+		{expectCanonicalInt: true, maxIntBytes: maxMathOpCodeLen},
```

Comment:
Maybe totally unrelated, or dex specific, but in our dex comments, we have this:
```
	// 5 bytes if-else-endif-equalverify-checksig
	// 1 + 1 + 1 + 1 + 1 + 1 + 32 + 1 + 1 + 1 + 1 + 20 = 62 bytes for redeem block
	// 1 + 4 + 1 + 1 + 1 + 1 + 1 + 20 = 30 bytes for refund block
	// 5 + 62 + 30 = 97 bytes
	//
	// Note that this allows for a secret size of up to 75 bytes, but the secret
	// must be 32 bytes to be considered valid.
```

Not sure why only 75 bytes. Does `maxIntBytes: maxMathOpCodeLen` allow for 0xFFFF byte long secrets since `txscript.MathOpCodeMaxScriptNumLen` is 4 (I think)? Would there be a reason secret length can't exceed 75?

Reply:
I think the dex parsing is more constrained, requiring `OP_DATA_1 secretSize` as the encoding for the `<secretSize>` data push, no `OP_PUSHDATA{1,2}`. That would restrict to specifying up to a 255 bytes secret with `OP_DATA_1` I believe.

So I suspect the 75 comment in the pkScript parsing dex code is a mistake, perhaps something to do with `OP_DATA_75` being the max `OP_DATA*`, but that would relate the sigScript with the actual secret data push, not the contract script.

Reply:
> Not sure why only 75 bytes. Does `maxIntBytes: maxMathOpCodeLen` allow for 0xFFFF byte long secrets since `txscript.MathOpCodeMaxScriptNumLen` is 4 (I think)? Would there be a reason secret length can't exceed 75?

Not sure about the 75 bytes comment in the DEX code, but it is most certainly incorrect.  I can hazard a guess that maybe it was stated that way through confusion with `OP_DATA_75` being the largest possible single-byte opcode for pushing data, but that doesn't apply here.

This technically allows the size to be up to `2^31 - 1`, since that is the maximum script number that can be canonically encoded with 4 bytes.


> The dex's parsing is more constrained, requiring `OP_DATA_1 secretSize` as the encoding for the `<secretSize>` data push, no `OP_PUSHDATA{1,2}`. That would restrict to specifying size up to a 255 bytes with `OP_DATA_1` I believe. 


Looking for `OP_DATA_1` would cause it to miss sizes 16 and under since those are explicitly the small integer opcodes when encoded canonically (e.g. `OP_16` instead of `OP_DATA_1 16`.  If you are looking explicitly for 32, then it would be fine, but this template code is intended to be a little less restrictive.

`OP_DATA_1` would indeed allow a size value of up to 255, inclusive, to be specified.

Reply:
>  If you are looking explicitly for 32, then it would be fine, but this template code is intended to be a little less restrictive

Yup, that's the case.  It has to be 32 in DEX, so we constrained the script structure more.

I'm pretty sure the 75 comment stemmed from musing on the input script in the redeem, hence out of place.

Reply:
I'm not horribly opposed to making this explicitly look for 32-bytes either, but I figured it would be more flexible for callers to just be able to do the following:
```Go
data := ExtractAtomicSwapDataPushesV0(redeemScript)
if data.SecretSize != 32 { /* not one we care about */ }
```

That way they can trivially move to other hash functions if the need should ever arise.

All of that said, I generally think it's probably better for things that work directly with complex redeem scripts like these to control their own destiny, so to speak, and perform their own detection.  I almost didn't even add this code in as part of this PR for that reason, but several people I spoke to independently said they preferred to have it available via a well tested and maintained package if for no other reason than acting as a jumping off point.